Remove `type_unpacks`

ncik-roberts opened the pull request:
The branch targeted by this PR refactored `type_unpacks` to do very little: just add some modules to the typing environment, and typecheck some expression, using `type_expect`, in that context. This PR simplifies the setup to remove `type_unpacks`, and just inline `type_expect` at the various callsites.

This means that we have to arrange a different way for modules to be added to the typing context. I separate this out to a different step, `add_module_variables`. We can't have `type_pattern_list` call `add_module_variables` because there are occasions where we need to typecheck some other expression and learn that a module variable indeed has a packed module type before `add_module_variables` can succeed. For instance:

```ocaml
let m : (module S) = ... in
let (module M) = m in
...
```

Knowing that the pattern variable associated with `M` has a packed module type relies on first inferring a type for `m`.

There are a few smaller things that are made nicer by this patch:
  * It's less awkward to return the extended environment back to the `type_let`. We need to do this so `type_let` can check scope escape under the appropriate environment.
  * It's easier to pass in an outer level into `add_module_variables` so we can call `Mtype.lower_nongen outer_level modl.mod_type;`, which is less awkward than calling it with 0. (See the attached comment.)

ncik-roberts committed:
Remove type_unpacks

ncik-roberts committed:
Comment the delayed call to `add_module_variables`

goldfirere commented:
There seem to be unrelated commits in this PR. Could you maybe rebase / change the settings in this PR to fix this? If you'd like help figuring out what to do, happy to puzzle it out with you.

ncik-roberts commented:
> There seem to be unrelated commits in this PR. Could you maybe rebase / change the settings in this PR to fix this? If you'd like help figuring out what to do, happy to puzzle it out with you.

Should be fixed now.

goldfirere commented on typing/typecore.ml:
```diff
@@ -744,6 +750,40 @@ let add_pattern_variables ?check ?check_as env pv =
+                  (mkloc (Longident.Lident unpack.tu_name.txt)
+                    unpack.tu_name.loc)))
+      in
+      Mtype.lower_nongen outer_level modl.mod_type;
```
This has crept back in, presumably accidentally.

ncik-roberts replied:
Yes, a mistake. It's gone now.

goldfirere commented on typing/typecore.ml:
```diff
@@ -2602,12 +2642,7 @@ let type_pattern_list
-       tu_uid = Uid.mk ~current_unit:(Env.get_unit_name ())}
-    ) (get_ref module_variables)
-  in
+  let unpacks = module_variables_to_unpacks (get_ref module_variables) in
   let new_env = add_pattern_variables !new_env pvs in
```
```suggestion
  let unpacks = module_variables_to_unpacks (get_ref module_variables) in
  let new_env = add_pattern_variables !new_env pvs in
  (* Don't call add_module_variables here, because its use of [type_module]
     will fail until after we have type-checked the expressions of the let.
     Example: [let (m : (module S)) = ... in let (module M) = m in ...]
     We learn the signature [S] from the type of [m] in the RHS of the second
     let, and we need that knowledge for [type_module} to succeed.
     If we type-checked expressions before patterns, then we could call
     [add_module_variables] here.
     *)
```

Maybe that's a bit excessive, but I think both you and I thought of adding `add_module_variables` here, and so someone else might wonder why we didn't.

ncik-roberts replied:
I lean on the side of adding a similarly detailed comment, given the opportunity for confusion here. I adopted this comment (almost verbatim). I found it clearer to move the responsibility of calling `add_pattern_variables` also outside of `type_pattern_list`, so I put this comment at the sole callsite of that function.

goldfirere commented on typing/typecore.ml:
```diff
@@ -6911,7 +6879,13 @@ and type_let
             check_partial_application ~statement:false vb_expr
       | _ -> ()) l;
-  (l, new_env, unpacks)
+  (* Note that we take care to avoid calling [add_module_variables]
```
Ah. You've already written the comment I wanted. But it needs to be in two places. I am a huge fan of GHC's `Note` convention, where this might become

```
Note [add_module_variables after checking expressions]
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
<< text here >>
```

and then both here and in `type_pattern_list`, there is `See Note [add_module_variables after checking expressions]`.

ncik-roberts replied:
See above: I adopted something like this convention, but the note is localized to `type_cases`. (Did you have another place in mind for the parent note?)

goldfirere reviewed (commented):
:tada:

ncik-roberts committed:
Excise `outer_level`

ncik-roberts committed:
Unify terminology for module/pattern variables

In the parent PR we unified the `to_unpack` and `module_variable` types.  This left something to be desired, as the `module_variable` terminology remained in some places. This PR removes the `to_unpack` terminology in favor of calling it `module_variable`; this makes sense with the joint removal of `type_unpacks` and leads to parallel code to `pattern_variable`.

ncik-roberts commented:
(Some tests are failing, I'm looking.)

ncik-roberts committed:
Comment the order of typing lets that bind modules

It's natural to want to call `add_module_variables`
within `type_pattern_list`; this comment clarifies why
we don't do that.

I found it clearer to also move the responsibility to call
`add_pattern_variables` outside of `type_pattern_list`, so
I make that change as well.
